File: src/qmd-rust/src/cli/embed.rs
use crate::cli::EmbedArgs;
use crate::store::Store;
use crate::store::chunker::{chunk_document, DEFAULT_CHUNK_SIZE, DEFAULT_OVERLAP};
use crate::llm::Router;
use anyhow::Result;

/// Handle embed command - generate/update embeddings
pub fn handle(
    cmd: &EmbedArgs,
    store: &Store,
    llm: &Router,
) -> Result<()> {
    let collection = cmd.collection.clone();

    // Create a Tokio runtime for async operations
    let rt = tokio::runtime::Runtime::new()?;

    if let Some(col) = &collection {
        println!("Generating embeddings for collection: {}", col);
        rt.block_on(async {
            embed_collection_async(store, col, llm, cmd.force).await
        })?;
    } else {
        println!("Generating embeddings for all collections...");
        rt.block_on(async {
            embed_all_collections_async(store, llm, cmd.force).await
        })?;
    }

    Ok(())
}

async fn embed_collection_async(
    store: &Store,
    collection: &str,
    llm: &Router,
    force: bool,
) -> Result<()> {
    use log::info;

    info!("Embedding collection: {}", collection);

    if !llm.has_embedder() {
        log::warn!("No embedder available, skipping embedding");
        return Ok(());
    }

    let conn = store.get_connection(collection)?;

    // Get all documents that need embedding
    let mut stmt = if force {
        conn.prepare("SELECT id, hash, doc FROM documents WHERE active = 1")?
    } else {
        conn.prepare(
            "SELECT id, hash, doc FROM documents
             WHERE active = 1
             AND hash NOT IN (SELECT DISTINCT hash FROM content_vectors)"
        )?
    };

    let docs: Vec<(i64, String, String)> = stmt
        .query_map([], |row| {
            Ok((row.get(0)?, row.get(1)?, row.get(2)?))
        })?
        .filter_map(|r| r.ok())
        .collect();

    info!("Found {} documents to embed", docs.len());

    // Chunk all documents and build a flat list of (hash, chunk) pairs
    let mut all_chunks: Vec<(String, crate::store::chunker::Chunk)> = Vec::new();

    for (_doc_id, hash, doc) in &docs {
        let chunks = chunk_document(doc, DEFAULT_CHUNK_SIZE, DEFAULT_OVERLAP);

        // If force re-embed, delete old chunks for this hash
        if force {
            conn.execute("DELETE FROM content_vectors WHERE hash = ?", [hash])?;
            // Delete all vector entries for this hash (any seq)
            let like_pattern = format!("{}_%", hash);
            conn.execute("DELETE FROM vectors_vec WHERE hash_seq LIKE ?", [&like_pattern])?;
        }

        for chunk in chunks {
            all_chunks.push((hash.clone(), chunk));
        }
    }

    info!("Total chunks to embed: {}", all_chunks.len());

    // Process chunks in batches
    let batch_size = 10;
    for (batch_idx, batch) in all_chunks.chunks(batch_size).enumerate() {
        info!("Processing batch {}/{}", batch_idx + 1, (all_chunks.len() + batch_size - 1) / batch_size);

        // Prepare texts for embedding
        let texts: Vec<&str> = batch.iter().map(|(_, chunk)| chunk.text.as_str()).collect();

        // Generate embeddings
        let embedding_result = llm.embed(&texts).await?;

        info!("Generated {} embeddings with model: {}",
              embedding_result.embeddings.len(), embedding_result.model);

        // Store embeddings per chunk
        for (i, (hash, chunk)) in batch.iter().enumerate() {
            let embedding = &embedding_result.embeddings[i];
            let embedding_json = serde_json::to_string(embedding)?;

            // Store in content_vectors metadata table
            conn.execute(
                "INSERT OR REPLACE INTO content_vectors (hash, seq, pos, model, embedded_at)
                 VALUES (?, ?, ?, ?, datetime('now'))",
                rusqlite::params![hash, chunk.seq as i64, chunk.pos as i64, &embedding_result.model],
            )?;

            // Store in vectors_vec table
            let hash_seq = format!("{}_{}", hash, chunk.seq);
            conn.execute(
                "INSERT OR REPLACE INTO vectors_vec (hash_seq, embedding)
                 VALUES (?, ?)",
                [&hash_seq, &embedding_json],
            )?;

            info!("Stored embedding for hash: {} chunk seq: {}", hash, chunk.seq);
        }
    }

    info!("Embedding complete for collection: {}", collection);
    Ok(())
}

async fn embed_all_collections_async(
    store: &Store,
    llm: &Router,
    force: bool,
) -> Result<()> {
    for collection in store.get_collections() {
        embed_collection_async(store, &collection.name, llm, force).await?;
    }
    Ok(())
}

File: qmd-go/pkg/cli/embed.go
package cli

import (
	"fmt"
	"os"

	"github.com/qmd/qmd-go/pkg/config"
	"github.com/qmd/qmd-go/pkg/llm"
	"github.com/qmd/qmd-go/pkg/store"
	"github.com/spf13/cobra"
)

var embedCmd = &cobra.Command{
	Use:   "embed [collection]",
	Short: "Generate/update embeddings for documents",
	Long:  `Generate or update vector embeddings for documents in a collection.`,
	Args:  cobra.MinimumNArgs(0),
	Run: func(cmd *cobra.Command, args []string) {
		collection := ""
		if len(args) > 0 {
			collection = args[0]
		}

		force, _ := cmd.Flags().GetBool("force")

		if err := runEmbed(collection, force); err != nil {
			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
			os.Exit(1)
		}
	},
}

func init() {
	embedCmd.Flags().BoolP("force", "f", false, "Force regeneration of all embeddings")
}

func runEmbed(collection string, force bool) error {
	// Load config
	cfg, err := config.Load("")
	if err != nil {
		return fmt.Errorf("failed to load config: %w", err)
	}

	// Get store
	s, err := getStore()
	if err != nil {
		return fmt.Errorf("failed to open store: %w", err)
	}
	defer s.Close()

	// Create LLM router
	router := llm.New(cfg)

	// Initialize embedder from config
	if cfg.Models.Embed != nil && cfg.Models.Embed.Local != "" {
		embedder, err := llm.NewLocalEmbedder(cfg.Models.Embed.Local)
		if err != nil {
			fmt.Printf("Warning: Failed to load embedder: %v\n", err)
			fmt.Println("Using random embeddings for demonstration...")
		} else {
			router.SetLocalEmbedder(embedder)
		}
	}

	if collection != "" {
		return embedCollection(s, router, collection, force)
	}

	// Embed all collections
	collections, err := s.GetCollections()
	if err != nil {
		return fmt.Errorf("failed to get collections: %w", err)
	}

	for _, col := range collections {
		if err := embedCollection(s, router, col.Name, force); err != nil {
			fmt.Fprintf(os.Stderr, "Warning: Failed to embed collection %s: %v\n", col.Name, err)
		}
	}

	return nil
}

func embedCollection(s *store.Store, router *llm.Router, collection string, force bool) error {
	fmt.Printf("Generating embeddings for collection: %s\n", collection)

	// Get documents needing embedding
	docs, err := s.GetDocumentsForEmbedding(collection, force)
	if err != nil {
		return fmt.Errorf("failed to get documents: %w", err)
	}

	if len(docs) == 0 {
		fmt.Println("No documents need embedding")
		return nil
	}

	fmt.Printf("Found %d documents to embed\n", len(docs))

	// Delete existing embeddings if force mode
	if force {
		for _, doc := range docs {
			if err := s.DeleteEmbeddings(doc.Hash); err != nil {
				return fmt.Errorf("failed to delete old embeddings: %w", err)
			}
		}
	}

	// Determine embedding model
	modelName := "nomic-embed-text-v1.5"
	if router.HasEmbedder() {
		fmt.Printf("Using embedder: %s\n", modelName)
	}

	// Process documents in batches
	batchSize := 10
	processed := 0

	for i := 0; i < len(docs); i += batchSize {
		end := i + batchSize
		if end > len(docs) {
			end = len(docs)
		}

		batch := docs[i:end]
		fmt.Printf("Processing batch %d-%d...\n", i+1, end)

		// Chunk documents
		var texts []string
		var hashes []string
		var seqs []int
		var poss []int

		for _, doc := range batch {
			chunks := store.ChunkDocument(doc.Doc, store.DefaultChunkSize, store.DefaultOverlap)
			for _, chunk := range chunks {
				texts = append(texts, chunk.Text)
				hashes = append(hashes, doc.Hash)
				seqs = append(seqs, chunk.Seq)
				poss = append(poss, chunk.Pos)
			}
		}

		if len(texts) == 0 {
			continue
		}

		// Generate embeddings
		var embeddings [][]float32

		if router.HasEmbedder() {
			result, err := router.Embed(texts)
			if err != nil {
				fmt.Printf("Warning: Failed to generate embeddings: %v\n", err)
				// Fall back to random embeddings
				for range texts {
					embeddings = append(embeddings, randomEmbedding(768))
				}
			} else {
				embeddings = result.Embeddings
				if result.Model != "" {
					modelName = result.Model
				}
			}
		} else {
			// Use random embeddings as placeholder
			for range texts {
				embeddings = append(embeddings, randomEmbedding(768))
			}
			fmt.Println("Note: Using random embeddings (no embedder configured)")
		}

		// Store embeddings
		for j, embedding := range embeddings {
			err := s.StoreEmbedding(hashes[j], seqs[j], poss[j], embedding, modelName)
			if err != nil {
				return fmt.Errorf("failed to store embedding: %w", err)
			}
		}

		processed += len(batch)
		fmt.Printf("Embedded %d/%d documents\n", processed, len(docs))
	}

	fmt.Println("Embedding complete!")
	return nil
}

// randomEmbedding generates a random embedding vector
func randomEmbedding(dim int) []float32 {
	embedding := make([]float32, dim)
	// Simple random initialization (not normalized)
	for i := range embedding {
		embedding[i] = float32(i%100) / 100.0
	}
	// Normalize
	var sum float32
	for _, v := range embedding {
		sum += v * v
	}
	norm := float32(0)
	if sum > 0 {
		norm = 1.0 / float32(sqrt(float64(sum)))
	}
	for i := range embedding {
		embedding[i] *= norm
	}
	return embedding
}

func sqrt(x float64) float64 {
	if x == 0 {
		return 0
	}
	// Simple approximation
	r := x
	for i := 0; i < 20; i++ {
		r = (r + x/r) / 2
	}
	return r
}
